### PR TITLE
Add setting to disable Google Hangouts extension

### LINKED ISSUE #8257: QtWebEngine includes a hidden extension, available only to `*.google.com` domains
**Version info**:

qutebrowser v3.2.1
Git commit: 
Backend: QtWebEngine 6.7.2
  based on Chromium 118.0.5993.220
  with security patches up to 125.0.6422.142 (plus any distribution patches)
  (source: api)
Qt: 6.7.2

**Does the bug happen if you start with `--temp-basedir`?**:
Yes

**Description**
It came to my attention that [Chrome includes a hidden extension](https://simonwillison.net/2024/Jul/9/hangout_servicesthunkjs/), which gives Google domains exclusive access to additional system information.

This extension appears to also be included in qutebrowser.

**How to reproduce**

Open `google.com` and enter the following in the JS console:
```
chrome.runtime.sendMessage(
  "nkeimhogjdpnpccoofpliimaahmaaome",
  { method: "cpu.getInfo" },
  (response) => {
    console.log(JSON.stringify(response, null, 2));
  },
);
```
You will get some output. On a non-google domain, you will instead get a `TypeError`.

This comes from Chromium via QtWebEngine, but I'm not sure how to write a patch to disable or remove it. Would be happy to do so with guidance.

### PR DESCRIPTION
**Pull request by illfygli**
Fixes #8257.

Hopefully, the extension can eventually be removed from QtWebEngine.

If not, and until then, this setting disables it, using the existing code in `pakjoy.py`.

### CODE REVIEW

File: doc/changelog.asciidoc [hunk 1]
```diff
@@ -19,6 +19,12 @@ breaking changes (such as renamed commands) can happen in minor releases.
 v3.3.0 (unreleased)
 -------------------
 
+Added
+~~~~~
+
+- A setting for disabling the Google Hangouts extension,
+  built into Chromium/QtWebEngine.
```
Comment (toofar) [inline doc/changelog.asciidoc:26]: nit: maybe mention the setting name here, eg "added the `qt.workarounds.disable_hangouts_extension` setting ..."

**Review by toofar — APPROVED** [review]
Seems fine to me! I appreciate that you went to the effort of updating the changelog and regenerating the docs.

I had a think about downsides (it's opt in and we made it pretty robust in the end) or if it could be made more generic for other builtin extensions (the extension loading code is very fragile, it just hard crashes with everything generic I could think of).

I won't merge right away to give @The-Compiler a chance to have a good ponder.

**Comment by The-Compiler** [comment]
This seems quite comprehensive, thanks!

A couple of things come to mind:

- It seems a bit confusing from an user's perspective that on certain Qt versions, the workaround still is enabled even if this is set to false. That probably makes sense since it'll lead to crashes on Google domains if we don't do that, but it's not clear to the user. Maybe we should at least add something to the settings description that the workaround is always active and the setting is ignored on:
  - Qt 6.5.0 to 6.5.3 if dark mode is enabled
  - Qt 6.6.0
- Alternatively we could make it an `always` / `auto` / `never` setting like some others, but I suppose the `QUTE_DISABLE_PAKJOY` environment variable is enough of an escape hatch.
- I usually try to avoid double negations in setting names (a `thing` setting with `false` by default is less confusing than a `disable_thing` setting with `true` by default). That being said, we have `qt.workarounds.disable_accelerated_2d_canvas` already, and in the scope of `qt.workarounds` (where I agree this belongs in), `disable_hangouts_extension` makes more sense indeed.
- Now that the workaround can be used with other Qt versions, we should probably aim to update `HANGOUTS_IDS` in `pakjoy.py` with the hardcoded IDs from other supported Qt versions. Doesn't necessary be part of this PR though.

None of this should be blocking IMHO, just some things that came to mind.

**Comment by illfygli** [comment]
@The-Compiler thanks for doing that, looks good to me.

Agreed about the double negatives too, and yeah, it just seemed nice to be consistent with those particular workaround settings, where that default makes sense.

**Comment by toofar** [comment]
:heavy_check_mark: Still looks good to me! Thanks for removing another grammar issue :)

**Comment by The-Compiler** [comment]
There's one more thing that irks me about this which I didn't consider before: So far, we treated this as a "best effort" patching, failing with just a stdout output if anything went wrong.

Given the user now explicitly requests it, and it's also something privacy sensitive, this doesn't seem like a sane failure mode anymore - if the user requested the patching, we should probably show an error message of some sorts when that was going wrong.

Here's an attempt at a fix:

```diff
diff --git i/qutebrowser/misc/pakjoy.py w/qutebrowser/misc/pakjoy.py
index f0c7d75b7..114bd2f72 100644
--- i/qutebrowser/misc/pakjoy.py
+++ w/qutebrowser/misc/pakjoy.py
@@ -34,7 +34,7 @@
 
 from qutebrowser.config import config
 from qutebrowser.misc import binparsing, objects
-from qutebrowser.utils import qtutils, standarddir, version, utils, log
+from qutebrowser.utils import qtutils, standarddir, version, utils, log, message
 
 HANGOUTS_MARKER = b"// Extension ID: nkeimhogjdpnpccoofpliimaahmaaome"
 HANGOUTS_IDS = [
@@ -234,7 +234,8 @@ def copy_webengine_resources() -> Optional[pathlib.Path]:
 def _patch(file_to_patch: pathlib.Path) -> None:
     """Apply any patches to the given pak file."""
     if not file_to_patch.exists():
-        log.misc.error(
+        _error(
+            None,
             "Resource pak doesn't exist at expected location! "
             f"Not applying quirks. Expected location: {file_to_patch}"
         )
@@ -247,8 +248,22 @@ def _patch(file_to_patch: pathlib.Path) -> None:
             offset = parser.find_patch_offset()
             binparsing.safe_seek(f, offset)
             f.write(REPLACEMENT_URL)
-        except binparsing.ParseError:
-            log.misc.exception("Failed to apply quirk to resources pak.")
+        except binparsing.ParseError as e:
+            _error(e, "Failed to apply quirk to resources pak.")
+
+
+def _error(exc: Optional[BaseException], text: str) -> None:
+    if config.val.qt.workarounds.disable_hangouts_extension:
+        # Explicitly requested -> hard error
+        lines = ["Failed to disable Hangouts extension"]
+        if exc is None:
+            lines.append(str(exc))
+        message.error("\n".join(lines))
+    elif exc is None:
+        # Best effort -> just log
+        log.misc.error(text)
+    else:
+        log.misc.exception(text)
 
 
 @contextlib.contextmanager
@@ -263,8 +278,8 @@ def patch_webengine() -> Iterator[None]:
         # Still calling this on Qt != 6.6 so that the directory is cleaned up
         # when not needed anymore.
         webengine_resources_path = copy_webengine_resources()
-    except OSError:
-        log.misc.exception("Failed to copy webengine resources, not applying quirk")
+    except OSError as e:
+        _error(e, "Failed to copy webengine resources, not applying quirk")
         yield
         return
 
```

but I haven't gotten around to testing this yet. Leaving for two weeks of holidays tomorrow morning, so not sure if I'll get around finishing it before that...

**Comment by illfygli** [comment]
@The-Compiler That's a good point!

I can try the patch out/modify if needed, and add it to the PR.

Enjoy your vacation!

**Comment by illfygli** [comment]
Tried it out. I get the error message in the GUI when it fails, so that seems good!

**Comment by The-Compiler** [comment]
Thanks! I added dfcb6e3f7c91f5a618dac195220a558c23517fb5 to improve the error message and adjust an automated test to check for it as well. Then (arguably a bit as an afterthought) I also pushed 6e84639276c70b14d76d0e46046b57adf6a68a34 to add some tests ensuring that explicit enabling works fine.

Thanks @illfygli for the PR and all the quick reactions, and @toofar for the reviewing! :sparkles:

**Pull request merged**